azurelinuxagent/ga/update.py

```diff
@@ -0,0 +1,278 @@
+import azurelinuxagent.common.logger as logger
+import azurelinuxagent.common.conf as conf
+from azurelinuxagent.common.event import add_event
+from azurelinuxagent.common.future import ustr
+import azurelinuxagent.common.utils.fileutil as fileutil
+import azurelinuxagent.common.utils.restutil as restutil
+from azurelinuxagent.common.version import AGENT_VERSION
+from azurelinuxagent.common.utils.textutil import Version
+from azurelinuxagent.common.osutil import get_osutil
+from azurelinuxagent.common.protocol import get_protocol_util
+
+def get_update_handler():
+    return UpdateHandler()
+
+GA_DIR = 'GuestAgent' # Dir name for guest agent bin and data
+GA_ERR = 'error' # File name for guest agent error record
+MAX_FAILURE = 3 # Max failure allowed for guest agent before blacklisted
+RETAIN_INTERVAL = 24 * 60 * 60 # Retain interval for black list
+
+"""
+Handles self update logic
+"""
+class UpdateHandler(object):
+
+    def __init__(self):
+        self.osutil = get_osutil()
+        self.protocol_util = get_protocol_util()
+        self.last_etag = None
+        self.agents = []
+        self.error_record = {}
+
+    def run(self):
+        """
+        1. If self-update is enabled, check for new versions.
+           otherwise, use current agent
+        2. Invoke run-exthandlers task
+        """
+        self.mk_ga_dir()
+
+        self.load_error_record() 
+
+        updated = self.check_for_update()
+        if not updated:
+            return 
+
+        latest_agent = None
+
+        if conf.get_autoupdate_enabled():
+            logger.info("Auto update enabled")
+            latest_agent = self.get_latest_agent()
+
+        try:
+            self.run_extensions(latest_agent)
+        except UpdateError as e:
+            add_event(u"WALA", is_success=False, message=ustr(e))
+            if latest_agent is not None:
+                latest_agent.err.mark_failure()
+        
+        self.save_error_record()
+    
+    def mk_ga_dir(self):
+        dir_path = os.path.join(conf.get_lib_dir(), GA_DIR)
+        fileutil.mkdir(dir_path, mode=0o700)
+
+    def get_latest_agent(self):
+        available_agent = [agent for agent in self.agents \
+                           if not agent.err.is_blacklisted()]
+ 
+        return available_agent[0] if len(available_agent) >= 1 else None
+
+    def run_extensions(self, latest_agent):
+        agent_bin = sys.argv[0]
+
+        if latest_agent is not None:
+            logger.info(u"The latest guest agent version is : {0}", 
+                        latest_agent.version)
+            agent_bin = latest_agent.get_agent_bin()
+
+            if not latest_agent.is_downloaded():
+                try:
+                    latest_agent.download()
+                except Exception as e:
+                    raise UpdateError(u"Download failed", e)
+
+        devnull = open(os.devnull, 'w')
+        try:
+            child = subprocess.Popen([agent_bin, 'run-exthandlers'], 
```

Comment:
@ahmetalpbalkan daemon/main.py contains a check_pid method that looks for / exits if it finds an existing pidfile (whose value corresponds to a running process) and exits if so. I believe it could be more robust.